File: 230_-_Kth_Smallest_Element_in_a_BST.cs
/**
 * 230. Kth Smallest Element in a BST - Medium
 * Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the values of the nodes in the tree.
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int KthSmallest(TreeNode root, int k) {
        return KthSmallestIterative(root, k);
    }
    
    public int KthSmallestIterative(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode current = root;
        
        while (current != null || stack.Count > 0) {
            while (current != null) {
                stack.Push(current);
                current = current.left;
            }
            
            current = stack.Pop();
            
            --k;
            if (k == 0) {
                return current.val;
            }
            
            current = current.right;
        }
        
        return -1;
    }
}

File: javascript/230_-_Kth_Smallest_Element_in_a_BST.js
/**
 * 230. Kth Smallest Element in a BST - Medium
 * Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the values of the nodes in the tree.
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    return kthSmallestIterative(root, k);
};

var kthSmallestIterative = function(root, k) {
    const stack = [];
    let current = root;

    while (current || stack.length > 0) {
        while (current) {
            stack.push(current);
            current = current.left;
        }

        current = stack.pop();

        --k;
        if (k == 0) {
            return current.val;
        }

        current = current.right;
    }

    return -1;
};